### ISSUE TITLE
difficulty compiling with older (CppAD) autodiff library

### ISSUE THREAD

**Issue by bbolker** [glmmTMB/glmmTMB#970]
Currently getting compilation errors when commenting out  `PKG_CPPFLAGS = -DTMBAD_FRAMEWORK` and building with the current master branch (errors copied below).  Based on the error, I think it has something to do with commit d22bf6161a7, which added a `log1p()` into the expression for the log-Normal family (the only time `log1p` occurs in the source code). I guess the way `log1p` is implemented isn't backward compatible with the CppAD framework ... ??

---

```
using C++ compiler: ‘g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
g++ -std=gnu++17 -I"/usr/local/lib/R/include" -DNDEBUG  -I'/usr/local/lib/R/site-library/TMB/include' -I'/usr/local/lib/R/site-library/RcppEigen/include' -I/usr/local/include   -fopenmp -fpic  -g -O2  -Wno-unused-variable -Wno-unused-function -Wno-ignored-attributes -c glmmTMB.cpp -o glmmTMB.o
glmmTMB.cpp: In instantiation of ‘Type objective_function<Type>::operator()() [with Type = CppAD::AD<double>]’:
/usr/local/lib/R/site-library/TMB/include/tmb_core.hpp:1308:6:   required from here
glmmTMB.cpp:812:19: error: cannot convert ‘CppAD::AD<double>’ to ‘double’
  812 |         s1 = log1p(pow(phi(i)/mu(i), 2.0)); //log1p(x) = log(1 + x), log-scale var
      |              ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
      |                   |
      |                   CppAD::AD<double>
In file included from /usr/include/c++/11/cmath:45,
                 from /usr/include/c++/11/complex:44,
                 from /usr/local/lib/R/site-library/RcppEigen/include/Eigen/Core:96,
                 from /usr/local/lib/R/site-library/RcppEigen/include/Eigen/Dense:1,
                 from /usr/local/lib/R/site-library/TMB/include/TMB.hpp:88,
                 from glmmTMB.cpp:1:
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:122:1: note:   initializing argument 1 of ‘double log1p(double)’
  122 | __MATHCALL_VEC (log1p,, (_Mdouble_ __x));
      | ^
glmmTMB.cpp: In instantiation of ‘Type objective_function<Type>::operator()() [with Type = CppAD::AD<CppAD::AD<double> >]’:
/usr/local/lib/R/site-library/TMB/include/tmb_core.hpp:913:29:   required from ‘Type objective_function<Type>::evalUserTemplate() [with Type = CppAD::AD<CppAD::AD<double> >]’
/usr/local/lib/R/site-library/TMB/include/tmb_core.hpp:2112:26:   required from here
glmmTMB.cpp:812:19: error: cannot convert ‘CppAD::AD<CppAD::AD<double> >’ to ‘double’
  812 |         s1 = log1p(pow(phi(i)/mu(i), 2.0)); //log1p(x) = log(1 + x), log-scale var
      |              ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
      |                   |
      |                   CppAD::AD<CppAD::AD<double> >
In file included from /usr/include/c++/11/cmath:45,
                 from /usr/include/c++/11/complex:44,
                 from /usr/local/lib/R/site-library/RcppEigen/include/Eigen/Core:96,
                 from /usr/local/lib/R/site-library/RcppEigen/include/Eigen/Dense:1,
                 from /usr/local/lib/R/site-library/TMB/include/TMB.hpp:88,
                 from glmmTMB.cpp:1:
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:122:1: note:   initializing argument 1 of ‘double log1p(double)’
  122 | __MATHCALL_VEC (log1p,, (_Mdouble_ __x));
      | ^
glmmTMB.cpp: In instantiation of ‘Type objective_function<Type>::operator()() [with Type = CppAD::AD<CppAD::AD<CppAD::AD<double> > >]’:
/usr/local/lib/R/site-library/TMB/include/tmb_core.hpp:913:29:   required from ‘Type objective_function<Type>::evalUserTemplate() [with Type = CppAD::AD<CppAD::AD<CppAD::AD<double> > >]’
/usr/local/lib/R/site-library/TMB/include/tmb_core.hpp:2361:28:   required from here
glmmTMB.cpp:812:19: error: cannot convert ‘CppAD::AD<CppAD::AD<CppAD::AD<double> > >’ to ‘double’
  812 |         s1 = log1p(pow(phi(i)/mu(i), 2.0)); //log1p(x) = log(1 + x), log-scale var
      |              ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
      |                   |
      |                   CppAD::AD<CppAD::AD<CppAD::AD<double> > >
In file included from /usr/include/c++/11/cmath:45,
                 from /usr/include/c++/11/complex:44,
                 from /usr/local/lib/R/site-library/RcppEigen/include/Eigen/Core:96,
                 from /usr/local/lib/R/site-library/RcppEigen/include/Eigen/Dense:1,
                 from /usr/local/lib/R/site-library/TMB/include/TMB.hpp:88,
                 from glmmTMB.cpp:1:
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:122:1: note:   initializing argument 1 of ‘double log1p(double)’
  122 | __MATHCALL_VEC (log1p,, (_Mdouble_ __x));
      | ^
make: *** [/usr/local/lib/R/etc/Makeconf:200: glmmTMB.o] Error 1
ERROR: compilation failed for package ‘glmmTMB’
* removing ‘/usr/local/lib/R/site-library/glmmTMB’
* restoring previous ‘/usr/local/lib/R/site-library/glmmTMB’
```

**Comment by bbolker** [comment]
It looks like I can work around this by replacing `log1p` with `logspace_add(., Type(0))`, but it does raise the general question (for curiosity) of (1) why this works with `TMBad` and not `CppAD` and (2) whether it would be good to have convenience functions (e.g.) `log1p`, `log1m` that were back-compatible with `CppAD` ... ??

**Comment by kaskr** [comment]
Yes, better use `logspace_add` to be consistent with the rest of the code (and to retain CppAD compatibility).
The version of CppAD that TMB uses is very old. Back then `log1p` was not available.

**Comment by bbolker** [comment]
Fixed in [227ed80](https://github.com/glmmTMB/glmmTMB/commit/227ed805f59dacfd25f95a44c430e9d189dd8086)
